File: packages/firebase_common/lib/src/firestore_documents/job.flutterfire_gen.dart
// coverage:ignore-file
// ignore_for_file: type=lint
// ignore_for_file: unused_element, deprecated_member_use, deprecated_member_use_from_same_package, use_function_type_syntax_for_parameters, unnecessary_const, avoid_init_to_null, invalid_override_different_default_values_named, prefer_expression_function_bodies, annotate_overrides, invalid_annotation_target, unnecessary_question_mark

part of 'job.dart';

class ReadJob {
  const ReadJob({
    required this.jobId,
    required this.path,
    required this.hostId,
    required this.imageUrl,
    required this.title,
    required this.place,
    required this.content,
    required this.belongings,
    required this.reward,
    required this.accessDescription,
    required this.accessTypes,
    required this.comment,
    required this.createdAt,
    required this.updatedAt,
  });

  final String jobId;

  final String path;

  final String hostId;

  final String imageUrl;

  final String title;

  final String place;

  final String content;

  final String belongings;

  final String reward;

  final String accessDescription;

  final Set<AccessType> accessTypes;

  final String comment;

  final DateTime? createdAt;

  final DateTime? updatedAt;

  factory ReadJob._fromJson(Map<String, dynamic> json) {
    return ReadJob(
      jobId: json['jobId'] as String,
      path: json['path'] as String,
      hostId: json['hostId'] as String,
      imageUrl: json['imageUrl'] as String? ?? '',
      title: json['title'] as String? ?? '',
      place: json['place'] as String? ?? '',
      content: json['content'] as String? ?? '',
      belongings: json['belongings'] as String? ?? '',
      reward: json['reward'] as String? ?? '',
      accessDescription: json['accessDescription'] as String? ?? '',
      accessTypes: json['accessTypes'] == null
          ? const <AccessType>{}
          : _accessTypesConverter
              .fromJson(json['accessTypes'] as List<dynamic>?),
      comment: json['comment'] as String? ?? '',
      createdAt: (json['createdAt'] as Timestamp?)?.toDate(),
      updatedAt: (json['updatedAt'] as Timestamp?)?.toDate(),
    );
  }

  factory ReadJob.fromDocumentSnapshot(DocumentSnapshot ds) {
    final data = ds.data()! as Map<String, dynamic>;
    return ReadJob._fromJson(<String, dynamic>{
      ...data,
      'jobId': ds.id,
      'path': ds.reference.path,
    });
  }
}

class CreateJob {
  const CreateJob({
    required this.hostId,
    this.imageUrl = '',
    required this.title,
    required this.place,
    required this.content,
    required this.belongings,
    required this.reward,
    this.accessDescription = '',
    this.accessTypes = const <AccessType>{},
    this.comment = '',
  });

  final String hostId;
  final String imageUrl;
  final String title;
  final String place;
  final String content;
  final String belongings;
  final String reward;
  final String accessDescription;
  final Set<AccessType> accessTypes;
  final String comment;

  Map<String, dynamic> toJson() {
    return {
      'hostId': hostId,
      'imageUrl': imageUrl,
      'title': title,
      'place': place,
      'content': content,
      'belongings': belongings,
      'reward': reward,
      'accessDescription': accessDescription,
      'accessTypes': _accessTypesConverter.toJson(accessTypes),
      'comment': comment,
      'createdAt': FieldValue.serverTimestamp(),
      'updatedAt': FieldValue.serverTimestamp(),
    };
  }
}

class UpdateJob {
  const UpdateJob({
    this.hostId,
    this.imageUrl,
    this.title,
    this.place,
    this.content,
    this.belongings,
    this.reward,
    this.accessDescription,
    this.accessTypes,
    this.comment,
    this.createdAt,
  });

  final String? hostId;
  final String? imageUrl;
  final String? title;
  final String? place;
  final String? content;
  final String? belongings;
  final String? reward;
  final String? accessDescription;
  final Set<AccessType>? accessTypes;
  final String? comment;
  final DateTime? createdAt;

  Map<String, dynamic> toJson() {
    return {
      if (hostId != null) 'hostId': hostId,
      if (imageUrl != null) 'imageUrl': imageUrl,
      if (title != null) 'title': title,
      if (place != null) 'place': place,
      if (content != null) 'content': content,
      if (belongings != null) 'belongings': belongings,
      if (reward != null) 'reward': reward,
      if (accessDescription != null) 'accessDescription': accessDescription,
      if (accessTypes != null)
        'accessTypes': _accessTypesConverter.toJson(accessTypes!),
      if (comment != null) 'comment': comment,
      if (createdAt != null) 'createdAt': createdAt,
      'updatedAt': FieldValue.serverTimestamp(),
    };
  }
}

class DeleteJob {}

/// Provides a reference to the jobs collection for reading.
final readJobCollectionReference =
    FirebaseFirestore.instance.collection('jobs').withConverter<ReadJob>(
          fromFirestore: (ds, _) => ReadJob.fromDocumentSnapshot(ds),
          toFirestore: (_, __) => throw UnimplementedError(),
        );

/// Provides a reference to a job document for reading.
DocumentReference<ReadJob> readJobDocumentReference({
  required String jobId,
}) =>
    readJobCollectionReference.doc(jobId);

/// Provides a reference to the jobs collection for creating.
final createJobCollectionReference =
    FirebaseFirestore.instance.collection('jobs').withConverter<CreateJob>(
          fromFirestore: (_, __) => throw UnimplementedError(),
          toFirestore: (obj, _) => obj.toJson(),
        );

/// Provides a reference to a job document for creating.
DocumentReference<CreateJob> createJobDocumentReference({
  required String jobId,
}) =>
    createJobCollectionReference.doc(jobId);

/// Provides a reference to the jobs collection for updating.
final updateJobCollectionReference =
    FirebaseFirestore.instance.collection('jobs').withConverter<UpdateJob>(
          fromFirestore: (_, __) => throw UnimplementedError(),
          toFirestore: (obj, _) => obj.toJson(),
        );

/// Provides a reference to a job document for updating.
DocumentReference<UpdateJob> updateJobDocumentReference({
  required String jobId,
}) =>
    updateJobCollectionReference.doc(jobId);

/// Provides a reference to the jobs collection for deleting.
final deleteJobCollectionReference =
    FirebaseFirestore.instance.collection('jobs').withConverter<DeleteJob>(
          fromFirestore: (_, __) => throw UnimplementedError(),
          toFirestore: (_, __) => throw UnimplementedError(),
        );

/// Provides a reference to a job document for deleting.
DocumentReference<DeleteJob> deleteJobDocumentReference({
  required String jobId,
}) =>
    deleteJobCollectionReference.doc(jobId);

/// Manages queries against the jobs collection.
class JobQuery {
  /// Fetches [ReadJob] documents.
  Future<List<ReadJob>> fetchDocuments({
    GetOptions? options,
    Query<ReadJob>? Function(Query<ReadJob> query)? queryBuilder,
    int Function(ReadJob lhs, ReadJob rhs)? compare,
  }) async {
    Query<ReadJob> query = readJobCollectionReference;
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    final qs = await query.get(options);
    final result = qs.docs.map((qds) => qds.data()).toList();
    if (compare != null) {
      result.sort(compare);
    }
    return result;
  }

  /// Subscribes [Job] documents.
  Stream<List<ReadJob>> subscribeDocuments({
    Query<ReadJob>? Function(Query<ReadJob> query)? queryBuilder,
    int Function(ReadJob lhs, ReadJob rhs)? compare,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    Query<ReadJob> query = readJobCollectionReference;
    if (queryBuilder != null) {
      query = queryBuilder(query)!;
    }
    var streamQs =
        query.snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamQs = streamQs.where((qs) => !qs.metadata.hasPendingWrites);
    }
    return streamQs.map((qs) {
      final result = qs.docs.map((qds) => qds.data()).toList();
      if (compare != null) {
        result.sort(compare);
      }
      return result;
    });
  }

  /// Fetches a specific [ReadJob] document.
  Future<ReadJob?> fetchDocument({
    required String jobId,
    GetOptions? options,
  }) async {
    final ds = await readJobDocumentReference(
      jobId: jobId,
    ).get(options);
    return ds.data();
  }

  /// Subscribes a specific [Job] document.
  Stream<ReadJob?> subscribeDocument({
    required String jobId,
    bool includeMetadataChanges = false,
    bool excludePendingWrites = false,
  }) {
    var streamDs = readJobDocumentReference(
      jobId: jobId,
    ).snapshots(includeMetadataChanges: includeMetadataChanges);
    if (excludePendingWrites) {
      streamDs = streamDs.where((ds) => !ds.metadata.hasPendingWrites);
    }
    return streamDs.map((ds) => ds.data());
  }

  /// Adds a [Job] document.
  Future<DocumentReference<CreateJob>> add({
    required CreateJob createJob,
  }) =>
      createJobCollectionReference.add(createJob);

  /// Sets a [Job] document.
  Future<void> set({
    required String jobId,
    required CreateJob createJob,
    SetOptions? options,
  }) =>
      createJobDocumentReference(
        jobId: jobId,
      ).set(createJob, options);

  /// Updates a specific [Job] document.
  Future<void> update({
    required String jobId,
    required UpdateJob updateJob,
  }) =>
      updateJobDocumentReference(
        jobId: jobId,
      ).update(updateJob.toJson());

  /// Deletes a specific [Job] document.
  Future<void> delete({
    required String jobId,
  }) =>
      deleteJobDocumentReference(
        jobId: jobId,
      ).delete();
}

File: functions/src/firestore-documents/job.flutterfireGen.ts
import * as admin from 'firebase-admin'
import {
    DocumentData,
    DocumentReference,
    DocumentSnapshot,
    FieldValue,
    Query,
    QueryDocumentSnapshot,
    QuerySnapshot,
    SetOptions,
    Timestamp,
    WriteResult
} from 'firebase-admin/firestore'

export type AccessType =
    | `trainAvailable`
    | `busAvailable`
    | `parkingAvailable`
    | `walkableFromNearest`
    | `shuttleServiceAvailable`

export class ReadJob {
    constructor({
        jobId,
        path,
        hostId,
        imageUrl,
        title,
        place,
        content,
        belongings,
        reward,
        accessDescription,
        accessTypes,
        comment,
        createdAt,
        updatedAt
    }: {
        jobId: string
        path: string
        hostId: string
        imageUrl: string
        title: string
        place: string
        content: string
        belongings: string
        reward: string
        accessDescription: string
        accessTypes: Set<AccessType>
        comment: string
        createdAt?: Date
        updatedAt?: Date
    }) {
        this.jobId = jobId
        this.path = path
        this.hostId = hostId
        this.imageUrl = imageUrl
        this.title = title
        this.place = place
        this.content = content
        this.belongings = belongings
        this.reward = reward
        this.accessDescription = accessDescription
        this.accessTypes = accessTypes
        this.comment = comment
        this.createdAt = createdAt
        this.updatedAt = updatedAt
    }

    readonly jobId: string

    readonly path: string

    readonly hostId: string

    readonly imageUrl: string

    readonly title: string

    readonly place: string

    readonly content: string

    readonly belongings: string

    readonly reward: string

    readonly accessDescription: string

    readonly accessTypes: Set<AccessType>

    readonly comment: string

    readonly createdAt?: Date

    readonly updatedAt?: Date

    private static accessTypesConverterFromJson(
        accessTypes: unknown[] | undefined
    ): Set<AccessType> {
        return new Set((accessTypes ?? []).map((e) => e as AccessType))
    }

    private static fromJson(json: Record<string, unknown>): ReadJob {
        return new ReadJob({
            jobId: json[`jobId`] as string,
            path: json[`path`] as string,
            hostId: json[`hostId`] as string,
            imageUrl: (json[`imageUrl`] as string | undefined) ?? ``,
            title: (json[`title`] as string | undefined) ?? ``,
            place: (json[`place`] as string | undefined) ?? ``,
            content: (json[`content`] as string | undefined) ?? ``,
            belongings: (json[`belongings`] as string | undefined) ?? ``,
            reward: (json[`reward`] as string | undefined) ?? ``,
            accessDescription:
                (json[`accessDescription`] as string | undefined) ?? ``,
            accessTypes:
                json[`accessTypes`] == undefined
                    ? new Set()
                    : ReadJob.accessTypesConverterFromJson(
                          json[`accessTypes`] as unknown[]
                      ),
            comment: (json[`comment`] as string | undefined) ?? ``,
            createdAt: (json[`createdAt`] as Timestamp | undefined)?.toDate(),
            updatedAt: (json[`updatedAt`] as Timestamp | undefined)?.toDate()
        })
    }

    static fromDocumentSnapshot(ds: DocumentSnapshot): ReadJob {
        const data = ds.data()!
        const cleanedData: Record<string, unknown> = {}
        for (const [key, value] of Object.entries(data)) {
            cleanedData[key] = value === null ? undefined : value
        }
        return ReadJob.fromJson({
            ...cleanedData,
            jobId: ds.id,
            path: ds.ref.path
        })
    }
}

export class CreateJob {
    constructor({
        hostId,
        imageUrl,
        title,
        place,
        content,
        belongings,
        reward,
        accessDescription,
        accessTypes,
        comment
    }: {
        hostId: string
        imageUrl: string
        title: string
        place: string
        content: string
        belongings: string
        reward: string
        accessDescription: string
        accessTypes: Set<AccessType>
        comment: string
    }) {
        this.hostId = hostId
        this.imageUrl = imageUrl
        this.title = title
        this.place = place
        this.content = content
        this.belongings = belongings
        this.reward = reward
        this.accessDescription = accessDescription
        this.accessTypes = accessTypes
        this.comment = comment
    }

    readonly hostId: string

    readonly imageUrl: string

    readonly title: string

    readonly place: string

    readonly content: string

    readonly belongings: string

    readonly reward: string

    readonly accessDescription: string

    readonly accessTypes: Set<AccessType>

    readonly comment: string

    private static accessTypesConverterToJson(
        accessTypes: Set<AccessType>
    ): string[] {
        return [...accessTypes]
    }

    toJson(): Record<string, unknown> {
        return {
            hostId: this.hostId,
            imageUrl: this.imageUrl,
            title: this.title,
            place: this.place,
            content: this.content,
            belongings: this.belongings,
            reward: this.reward,
            accessDescription: this.accessDescription,
            accessTypes: CreateJob.accessTypesConverterToJson(this.accessTypes),
            comment: this.comment,
            createdAt: FieldValue.serverTimestamp(),
            updatedAt: FieldValue.serverTimestamp()
        }
    }
}

export class UpdateJob {
    constructor({
        hostId,
        imageUrl,
        title,
        place,
        content,
        belongings,
        reward,
        accessDescription,
        accessTypes,
        comment,
        createdAt
    }: {
        hostId?: string
        imageUrl?: string
        title?: string
        place?: string
        content?: string
        belongings?: string
        reward?: string
        accessDescription?: string
        accessTypes?: Set<AccessType>
        comment?: string
        createdAt?: Date
    }) {
        this.hostId = hostId
        this.imageUrl = imageUrl
        this.title = title
        this.place = place
        this.content = content
        this.belongings = belongings
        this.reward = reward
        this.accessDescription = accessDescription
        this.accessTypes = accessTypes
        this.comment = comment
        this.createdAt = createdAt
    }

    readonly hostId?: string

    readonly imageUrl?: string

    readonly title?: string

    readonly place?: string

    readonly content?: string

    readonly belongings?: string

    readonly reward?: string

    readonly accessDescription?: string

    readonly accessTypes?: Set<AccessType>

    readonly comment?: string

    readonly createdAt?: Date

    private static accessTypesConverterToJson(
        accessTypes: Set<AccessType>
    ): string[] {
        return [...accessTypes]
    }

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    toJson(): Record<string, any> {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const json: Record<string, any> = {}
        if (this.hostId != undefined) {
            json[`hostId`] = this.hostId
        }
        if (this.imageUrl != undefined) {
            json[`imageUrl`] = this.imageUrl
        }
        if (this.title != undefined) {
            json[`title`] = this.title
        }
        if (this.place != undefined) {
            json[`place`] = this.place
        }
        if (this.content != undefined) {
            json[`content`] = this.content
        }
        if (this.belongings != undefined) {
            json[`belongings`] = this.belongings
        }
        if (this.reward != undefined) {
            json[`reward`] = this.reward
        }
        if (this.accessDescription != undefined) {
            json[`accessDescription`] = this.accessDescription
        }
        if (this.accessTypes != undefined) {
            json[`accessTypes`] = UpdateJob.accessTypesConverterToJson(
                this.accessTypes
            )
        }
        if (this.comment != undefined) {
            json[`comment`] = this.comment
        }
        if (this.createdAt != undefined) {
            json[`createdAt`] = this.createdAt
        }
        json[`updatedAt`] = FieldValue.serverTimestamp()
        return json
    }
}

export class DeleteJob {}

/**
 * A Cloud Firestore object which ignores `undefined` properties.
 */
const db = admin.firestore()
db.settings({ ignoreUndefinedProperties: true })

/**
 * Provides a reference to the jobs collection for reading.
 */
export const readJobCollectionReference = db
    .collection(`jobs`)
    .withConverter<ReadJob>({
        fromFirestore: (ds: DocumentSnapshot): ReadJob => {
            return ReadJob.fromDocumentSnapshot(ds)
        },
        toFirestore: () => {
            throw new Error(`toFirestore is not implemented for ReadJob`)
        }
    })

/**
 * Provides a reference to a job document for reading.
 * @param jobId - The ID of the job document to read.
 */
export const readJobDocumentReference = ({
    jobId
}: {
    jobId: string
}): DocumentReference<ReadJob> => readJobCollectionReference.doc(jobId)

/**
 * Provides a reference to the jobs collection for creating.
 */
export const createJobCollectionReference = db
    .collection(`jobs`)
    .withConverter<CreateJob>({
        fromFirestore: () => {
            throw new Error(`fromFirestore is not implemented for CreateJob`)
        },
        toFirestore: (obj: CreateJob): DocumentData => {
            return obj.toJson()
        }
    })

/**
 * Provides a reference to a job document for creating.
 * @param jobId - The ID of the job document to read.
 */
export const createJobDocumentReference = ({
    jobId
}: {
    jobId: string
}): DocumentReference<CreateJob> => createJobCollectionReference.doc(jobId)

/**
 * Provides a reference to the jobs collection for updating.
 */
export const updateJobCollectionReference = db
    .collection(`jobs`)
    .withConverter<UpdateJob>({
        fromFirestore: () => {
            throw new Error(`fromFirestore is not implemented for UpdateJob`)
        },
        toFirestore: (obj: UpdateJob): DocumentData => {
            return obj.toJson()
        }
    })

/**
 * Provides a reference to a job document for updating.
 * @param jobId - The ID of the job document to read.
 */
export const updateJobDocumentReference = ({
    jobId
}: {
    jobId: string
}): DocumentReference<UpdateJob> => updateJobCollectionReference.doc(jobId)

/**
 * Provides a reference to the jobs collection for deleting.
 */
export const deleteJobCollectionReference = db
    .collection(`jobs`)
    .withConverter<DeleteJob>({
        fromFirestore: () => {
            throw new Error(`fromFirestore is not implemented for DeleteJob`)
        },
        toFirestore: (): DocumentData => {
            throw new Error(`toFirestore is not implemented for DeleteJob`)
        }
    })

/**
 * Provides a reference to a job document for deleting.
 * @param jobId - The ID of the job document to read.
 */
export const deleteJobDocumentReference = ({
    jobId
}: {
    jobId: string
}): DocumentReference<DeleteJob> => deleteJobCollectionReference.doc(jobId)

/**
 * Manages queries against the jobs collection.
 */
export class JobQuery {
    /**
     * Fetches job documents.
     * @param queryBuilder - Function to modify the query.
     * @param compare - Function to sort the results.
     */
    async fetchDocuments({
        queryBuilder,
        compare
    }: {
        queryBuilder?: (query: Query<ReadJob>) => Query<ReadJob>
        compare?: (lhs: ReadJob, rhs: ReadJob) => number
    }): Promise<ReadJob[]> {
        let query: Query<ReadJob> = readJobCollectionReference
        if (queryBuilder != undefined) {
            query = queryBuilder(query)
        }
        const qs: QuerySnapshot<ReadJob> = await query.get()
        let result = qs.docs.map((qds: QueryDocumentSnapshot<ReadJob>) =>
            qds.data()
        )
        if (compare != undefined) {
            result = result.sort(compare)
        }
        return result
    }

    /**
     * Fetches a specific job document.
     * @param jobId - The ID of the job document to fetch.
     */
    async fetchDocument({
        jobId
    }: {
        jobId: string
    }): Promise<ReadJob | undefined> {
        const ds = await readJobDocumentReference({
            jobId
        }).get()
        return ds.data()
    }

    /**
     * Adds a job document.
     * @param createJob - The job details to add.
     */
    async add({
        createJob
    }: {
        createJob: CreateJob
    }): Promise<DocumentReference<CreateJob>> {
        return createJobCollectionReference.add(createJob)
    }

    /**
     * Sets a job document.
     * @param jobId - The ID of the job document to set.
     * @param createJob - The job details to set.
     * @param options - Options for the set operation.
     */
    async set({
        jobId,
        createJob,
        options
    }: {
        jobId: string
        createJob: CreateJob
        options?: SetOptions
    }): Promise<WriteResult> {
        if (options == undefined) {
            return createJobDocumentReference({
                jobId
            }).set(createJob)
        } else {
            return createJobDocumentReference({
                jobId
            }).set(createJob, options)
        }
    }

    /**
     * Updates a specific job document.
     * @param jobId - The ID of the job document to update.
     * @param updateJob - The details for updating the job.
     */
    async update({
        jobId,
        updateJob
    }: {
        jobId: string
        updateJob: UpdateJob
    }): Promise<WriteResult> {
        return updateJobDocumentReference({
            jobId
        }).update(updateJob.toJson())
    }

    /**
     * Deletes a specific job document.
     * @param jobId - The ID of the job document to delete.
     */
    async delete({ jobId }: { jobId: string }): Promise<WriteResult> {
        return deleteJobDocumentReference({
            jobId
        }).delete()
    }
}
